### DISCUSSION TITLE
Use functions without registering

### DISCUSSION THREAD

**Discussion by GitHunter0** [pwwang/datar#169] [General]
As far as I tested, only numpy functions work without the need of registering, right?

I tried to register the function `irr()` but I could not make the example below work, what's the right method to do it please?

```
import pandas as pd
import datar.all as d
from datar.all import f
import pyxirr

df = pd.DataFrame({ 'y': [-2, 1, 1, 1] })
df >> d.mutate(z = pyxirr.irr(f.y))
```

In the future, will `datar` be able to use more functions out of the box like in Pandas, in this case `df.assign(z = lambda x: pyxirr.irr(x.y))`?

Thank you

**Comment by pwwang** [comment]
`f.y` is an `Expression` object, which is used to defer the evaluation of the function/expression. Without registering, the function doesn't know how to handle an `Expression` object.

**Reply by GitHunter0** [reply]
Yes, but how can I register the function above? I could not find a way to do it properly.

I tried this for example:
```
    from pipda import register_func
    import pyxirr
    
    @register_func
    def irr(x):
        return pyxirr.irr(x) 
```

PS: there are some pages not opening:
https://pwwang.github.io/datar/advanced/
https://pwwang.github.io/datar/reference-maps/datar

**Reply by pwwang** [reply]
Those links don't exist anymore. You need to refer to the lastest docs:

https://pwwang.github.io/datar/

For registering functions/verbs, you can find instructions here:

https://pwwang.github.io/pipda/functions/
https://pwwang.github.io/pipda/verbs/

In your case:

```python
from pipda import register_func
import pandas as pd
import datar.all as d
from datar.all import f
import pyxirr

irr = register_func(pyxirr.irr)

df = pd.DataFrame({ 'y': [-2, 1, 1, 1] })
df >> d.mutate(z = irr(f.y))
```

**Comment by GitHunter0** [comment]
I had also tried that earlier but it didn't work, I found out there was a typo in my code... sorry. Thanks again for the support

**Reply by pwwang** [reply]
Good to know. A linter is your friend to avoid that.

**Reply by GitHunter0** [reply]
> Good to know. A linter is your friend to avoid that.

Indeed it is my friend.

PS: in https://pwwang.github.io/datar/reference-maps/ALL/  The reference hyperlink of `Datar-specific verbs/functions` is sending to the broken link https://pwwang.github.io/datar/reference-maps/datar

**Reply by pwwang** [reply]
It's renamed to `other` and then `misc`, since 0.11. The doc needs to be updated.
